### PR TITLE
External metastores support for most HDInsight resources

### PR DESCRIPTION
**Pull request by kosinsky**
Depends On #6809 

PR https://github.com/terraform-providers/terraform-provider-azurerm/pull/6145 added support for external metastores for azurerm_hdinsight_hadoop_cluster. Using the same approach to expand support for:
* azurerm_hdinsight_hbase_cluster
* azurerm_hdinsight_interactive_query_cluster
* azurerm_hdinsight_kafka_cluster
* azurerm_hdinsight_spark_cluster
* azurerm_hdinsight_storm_cluster

Sample configuration:
```hcl
metastores {
    hive {
        server   = azurerm_sql_server.test.fully_qualified_domain_name
        database_name = azurerm_sql_database.test.name
        username      = azurerm_sql_server.test.administrator_login
        password      = azurerm_sql_server.test.administrator_login_password
    }
    oozie {
        server   = azurerm_sql_server.test.fully_qualified_domain_name
        database_name = azurerm_sql_database.test.name
        username      = azurerm_sql_server.test.administrator_login
        password      = azurerm_sql_server.test.administrator_login_password
    }
    ambari {
        server   = azurerm_sql_server.test.fully_qualified_domain_name
        database_name = azurerm_sql_database.test.name
        username      = azurerm_sql_server.test.administrator_login
        password      = azurerm_sql_server.test.administrator_login_password
   }
}
``` 

New tests:
```bash
make testacc TEST='./azurerm/internal/services/hdinsight/tests/' TESTARGS='-run=Metastore -parallel=2' TESTTIMEOUT='300m'                                           
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
TF_ACC=1 go test ./azurerm/internal/services/hdinsight/tests/ -v -run=Metastore -parallel=2 -timeout 300m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMHDInsightHadoopCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightHadoopCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightHadoopCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightHadoopCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightHadoopCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightHadoopCluster_updateMetastore
=== RUN   TestAccAzureRMHDInsightHBaseCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightHBaseCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightHBaseCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightHBaseCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightHBaseCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightHBaseCluster_updateMetastore
=== RUN   TestAccAzureRMHDInsightInteractiveQueryCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightInteractiveQueryCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightInteractiveQueryCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightInteractiveQueryCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightInteractiveQueryCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightInteractiveQueryCluster_updateMetastore
=== RUN   TestAccAzureRMHDInsightKafkaCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightKafkaCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightKafkaCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightKafkaCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightKafkaCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightKafkaCluster_updateMetastore
=== RUN   TestAccAzureRMHDInsightSparkCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightSparkCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightSparkCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightSparkCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightSparkCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightSparkCluster_updateMetastore
=== RUN   TestAccAzureRMHDInsightStormCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightStormCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightStormCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightStormCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightStormCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightStormCluster_updateMetastore
=== CONT  TestAccAzureRMHDInsightHadoopCluster_allMetastores
=== CONT  TestAccAzureRMHDInsightKafkaCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightKafkaCluster_hiveMetastore (1121.89s)
=== CONT  TestAccAzureRMHDInsightStormCluster_allMetastores
--- PASS: TestAccAzureRMHDInsightStormCluster_allMetastores (898.22s)
=== CONT  TestAccAzureRMHDInsightStormCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightHadoopCluster_allMetastores (2111.41s)
=== CONT  TestAccAzureRMHDInsightStormCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightStormCluster_hiveMetastore (889.58s)
=== CONT  TestAccAzureRMHDInsightHBaseCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightStormCluster_updateMetastore (1651.68s)
=== CONT  TestAccAzureRMHDInsightKafkaCluster_allMetastores
--- PASS: TestAccAzureRMHDInsightKafkaCluster_allMetastores (1065.09s)
=== CONT  TestAccAzureRMHDInsightInteractiveQueryCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightHBaseCluster_updateMetastore (2024.42s)
=== CONT  TestAccAzureRMHDInsightInteractiveQueryCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightInteractiveQueryCluster_hiveMetastore (1014.18s)
=== CONT  TestAccAzureRMHDInsightInteractiveQueryCluster_allMetastores
--- PASS: TestAccAzureRMHDInsightInteractiveQueryCluster_updateMetastore (2084.51s)
=== CONT  TestAccAzureRMHDInsightSparkCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightInteractiveQueryCluster_allMetastores (1081.28s)
=== CONT  TestAccAzureRMHDInsightHBaseCluster_allMetastores
--- PASS: TestAccAzureRMHDInsightSparkCluster_hiveMetastore (1182.48s)
=== CONT  TestAccAzureRMHDInsightHBaseCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightHBaseCluster_allMetastores (1008.87s)
=== CONT  TestAccAzureRMHDInsightSparkCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightHBaseCluster_hiveMetastore (1056.78s)
=== CONT  TestAccAzureRMHDInsightHadoopCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightSparkCluster_updateMetastore (1892.70s)
=== CONT  TestAccAzureRMHDInsightHadoopCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightHadoopCluster_updateMetastore (1772.36s)
=== CONT  TestAccAzureRMHDInsightKafkaCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightHadoopCluster_hiveMetastore (1075.86s)
=== CONT  TestAccAzureRMHDInsightSparkCluster_allMetastores
--- PASS: TestAccAzureRMHDInsightSparkCluster_allMetastores (1066.44s)
--- PASS: TestAccAzureRMHDInsightKafkaCluster_updateMetastore (2011.75s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hdinsight/tests     12845.822s
```

I skipped MLService/RServer because they a pretty unstable (every other time even basic test fails with `Code="BadRequest" Message="HDI Version'3.6' is not supported for clusterType 'MLServices' and componentVersion 'default'."`) and [scheduled to retire by the end of 2020](https://docs.microsoft.com/en-us/azure/hdinsight/r-server/r-server-overview):
```
The ML Services cluster type on HDInsight is supported only on HDInsight 3.6. HDInsight 3.6 is scheduled to retire on December 31, 2020.
```

Fixes https://github.com/terraform-providers/terraform-provider-azurerm/issues/5230 for most HDInsight resources

### CODE REVIEW

**Comment by kosinsky** [comment]
@jackofallops, could you take a look? This PR is applying https://github.com/terraform-providers/terraform-provider-azurerm/pull/6145 approach for other HDInsight resources.

**Comment by ross-p-smith** [comment]
Hi @jackofallops - I require this change for a project I am working on. Do you know a rough ETA so that I can plan my sprints? Many thanks for any help you can offer

File: azurerm/internal/services/hdinsight/hdinsight_hbase_cluster_resource.go [hunk 1]
```diff
@@ -139,7 +141,14 @@ func resourceArmHDInsightHBaseClusterCreate(d *schema.ResourceData, meta interfa
...
 	gatewayRaw := d.Get("gateway").([]interface{})
-	gateway := azure.ExpandHDInsightsConfigurations(gatewayRaw)
+	configurations := azure.ExpandHDInsightsConfigurations(gatewayRaw)
+
+	if metastoresRaw, ok := d.GetOkExists("metastores"); ok {
+		metastores := expandHDInsightsMetastore(metastoresRaw.([]interface{}))
+		for k, v := range metastores {
+			configurations[k] = v
+		}
+	}
```
Comment (jackofallops) [inline azurerm/internal/services/hdinsight/hdinsight_hbase_cluster_resource.go:151]: `GetOkExists` is only intended for bools, and is also being deprecated in the plugin SDK in future.  We're working through removing / update other occurrences at the moment.  The following should be in line with the rest of the changes (along with the change to `expandHDInsightsMetastore` in 6809, so I'll make this PR dependent on that one)

```suggestion
	metastoresRaw, ok := d.Get("metastores").([]interface{})
	metastores := expandHDInsightsMetastore(metastoresRaw)
		for k, v := range metastores {
			configurations[k] = v
		}
	}
```

File: azurerm/internal/services/hdinsight/hdinsight_interactive_query_cluster_resource.go [hunk 2]
```diff
@@ -139,7 +141,14 @@ func resourceArmHDInsightInteractiveQueryClusterCreate(d *schema.ResourceData, m
...
 	gatewayRaw := d.Get("gateway").([]interface{})
-	gateway := azure.ExpandHDInsightsConfigurations(gatewayRaw)
+	configurations := azure.ExpandHDInsightsConfigurations(gatewayRaw)
+
+	if metastoresRaw, ok := d.GetOkExists("metastores"); ok {
+		metastores := expandHDInsightsMetastore(metastoresRaw.([]interface{}))
+		for k, v := range metastores {
+			configurations[k] = v
+		}
+	}
```
Comment (jackofallops) [inline azurerm/internal/services/hdinsight/hdinsight_interactive_query_cluster_resource.go:151]: As above

File: azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go [hunk 3]
```diff
@@ -140,7 +142,14 @@ func resourceArmHDInsightKafkaClusterCreate(d *schema.ResourceData, meta interfa
...
 	gatewayRaw := d.Get("gateway").([]interface{})
-	gateway := azure.ExpandHDInsightsConfigurations(gatewayRaw)
+	configurations := azure.ExpandHDInsightsConfigurations(gatewayRaw)
+
+	if metastoresRaw, ok := d.GetOkExists("metastores"); ok {
+		metastores := expandHDInsightsMetastore(metastoresRaw.([]interface{}))
+		for k, v := range metastores {
+			configurations[k] = v
+		}
+	}
```
Comment (jackofallops) [inline azurerm/internal/services/hdinsight/hdinsight_kafka_cluster_resource.go:152]: As above

File: azurerm/internal/services/hdinsight/hdinsight_spark_cluster_resource.go [hunk 4]
```diff
@@ -139,7 +141,14 @@ func resourceArmHDInsightSparkClusterCreate(d *schema.ResourceData, meta interfa
...
 	gatewayRaw := d.Get("gateway").([]interface{})
-	gateway := azure.ExpandHDInsightsConfigurations(gatewayRaw)
+	configurations := azure.ExpandHDInsightsConfigurations(gatewayRaw)
+
+	if metastoresRaw, ok := d.GetOkExists("metastores"); ok {
+		metastores := expandHDInsightsMetastore(metastoresRaw.([]interface{}))
+		for k, v := range metastores {
+			configurations[k] = v
+		}
+	}
```
Comment (jackofallops) [inline azurerm/internal/services/hdinsight/hdinsight_spark_cluster_resource.go:151]: As above

File: azurerm/internal/services/hdinsight/hdinsight_storm_cluster_resource.go [hunk 5]
```diff
@@ -138,7 +140,14 @@ func resourceArmHDInsightStormClusterCreate(d *schema.ResourceData, meta interfa
...
 	gatewayRaw := d.Get("gateway").([]interface{})
-	gateway := azure.ExpandHDInsightsConfigurations(gatewayRaw)
+	configurations := azure.ExpandHDInsightsConfigurations(gatewayRaw)
+
+	if metastoresRaw, ok := d.GetOkExists("metastores"); ok {
+		metastores := expandHDInsightsMetastore(metastoresRaw.([]interface{}))
+		for k, v := range metastores {
+			configurations[k] = v
+		}
+	}
```
Comment (jackofallops) [inline azurerm/internal/services/hdinsight/hdinsight_storm_cluster_resource.go:150]: As above

**Comment by kosinsky** [comment]
@jackofallops Done. 

make testacc TEST='./azurerm/internal/services/hdinsight/tests/' TESTARGS='-run=Metastore -parallel=2' TESTTIMEOUT='300m'
==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
TF_ACC=1 go test ./azurerm/internal/services/hdinsight/tests/ -v -run=Metastore -parallel=2 -timeout 300m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccAzureRMHDInsightHadoopCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightHadoopCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightHadoopCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightHadoopCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightHadoopCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightHadoopCluster_updateMetastore
=== RUN   TestAccAzureRMHDInsightHBaseCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightHBaseCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightHBaseCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightHBaseCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightHBaseCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightHBaseCluster_updateMetastore
=== RUN   TestAccAzureRMHDInsightInteractiveQueryCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightInteractiveQueryCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightInteractiveQueryCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightInteractiveQueryCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightInteractiveQueryCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightInteractiveQueryCluster_updateMetastore
=== RUN   TestAccAzureRMHDInsightKafkaCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightKafkaCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightKafkaCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightKafkaCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightKafkaCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightKafkaCluster_updateMetastore
=== RUN   TestAccAzureRMHDInsightSparkCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightSparkCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightSparkCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightSparkCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightSparkCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightSparkCluster_updateMetastore
=== RUN   TestAccAzureRMHDInsightStormCluster_allMetastores
=== PAUSE TestAccAzureRMHDInsightStormCluster_allMetastores
=== RUN   TestAccAzureRMHDInsightStormCluster_hiveMetastore
=== PAUSE TestAccAzureRMHDInsightStormCluster_hiveMetastore
=== RUN   TestAccAzureRMHDInsightStormCluster_updateMetastore
=== PAUSE TestAccAzureRMHDInsightStormCluster_updateMetastore
=== CONT  TestAccAzureRMHDInsightHadoopCluster_allMetastores
=== CONT  TestAccAzureRMHDInsightKafkaCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightHadoopCluster_allMetastores (1091.92s)
=== CONT  TestAccAzureRMHDInsightHBaseCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightKafkaCluster_hiveMetastore (1130.79s)
=== CONT  TestAccAzureRMHDInsightKafkaCluster_allMetastores
--- PASS: TestAccAzureRMHDInsightKafkaCluster_allMetastores (1131.90s)
=== CONT  TestAccAzureRMHDInsightInteractiveQueryCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightHBaseCluster_updateMetastore (2515.60s)
=== CONT  TestAccAzureRMHDInsightInteractiveQueryCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightInteractiveQueryCluster_updateMetastore (1969.68s)
=== CONT  TestAccAzureRMHDInsightInteractiveQueryCluster_allMetastores
--- PASS: TestAccAzureRMHDInsightInteractiveQueryCluster_hiveMetastore (1117.55s)
=== CONT  TestAccAzureRMHDInsightStormCluster_allMetastores
--- PASS: TestAccAzureRMHDInsightInteractiveQueryCluster_allMetastores (1278.22s)
=== CONT  TestAccAzureRMHDInsightStormCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightStormCluster_allMetastores (1043.68s)
=== CONT  TestAccAzureRMHDInsightStormCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightStormCluster_hiveMetastore (943.80s)
=== CONT  TestAccAzureRMHDInsightHBaseCluster_allMetastores
--- PASS: TestAccAzureRMHDInsightStormCluster_updateMetastore (1786.09s)
=== CONT  TestAccAzureRMHDInsightHBaseCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightHBaseCluster_allMetastores (1132.15s)
=== CONT  TestAccAzureRMHDInsightHadoopCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightHBaseCluster_hiveMetastore (1023.06s)
=== CONT  TestAccAzureRMHDInsightSparkCluster_allMetastores
--- PASS: TestAccAzureRMHDInsightSparkCluster_allMetastores (956.13s)
=== CONT  TestAccAzureRMHDInsightKafkaCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightHadoopCluster_updateMetastore (1748.23s)
=== CONT  TestAccAzureRMHDInsightSparkCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightSparkCluster_hiveMetastore (1012.98s)
=== CONT  TestAccAzureRMHDInsightSparkCluster_updateMetastore
--- PASS: TestAccAzureRMHDInsightKafkaCluster_updateMetastore (2028.44s)
=== CONT  TestAccAzureRMHDInsightHadoopCluster_hiveMetastore
--- PASS: TestAccAzureRMHDInsightHadoopCluster_hiveMetastore (1080.97s)
--- PASS: TestAccAzureRMHDInsightSparkCluster_updateMetastore (1965.47s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/hdinsight/tests     12573.878s

**Comment by tombuildsstuff** [comment]
👋 

Just to give an update here, whilst this PR was dependent on #6809 - we ended up disabling Binary Testing in that PR - work to enable Binary testing will happen in a subsequent PR (and is being investigated) - however unfortunately this PR is still blocked for the moment.

Thanks!

**Comment by kosinsky** [comment]
Thanks for letting me know. However, I'm not sure how it depends on on #6809. This PR is extenting of https://github.com/terraform-providers/terraform-provider-azurerm/pull/6145 to all HDI resources and #6145 was merged in 2.9 time frame.

**Comment by jackofallops** [comment]
@kosinsky - Apologies for the confusion, we had a crossed wire somewhere.  This PR was blocked by 6809, but for the `GetOkExists` deprecation, not the binary testing issue we're working on.  I've got this PR open now and working through it to resolve the merge conflict and get the testing kicked off on our CI server in preparation for merge.

Thanks again!

**Comment by jackofallops** [comment]
Hi @kosinsky - I've had to do a bit of git wrangling to account for a mistake we had to correct on master creating a bit of a diverge in history.  I think I've got it straight now, but if you can check everything looks correct to you I'll merge later (pending the tests on our CI, ofc)

Thanks!

**Comment by jackofallops** [comment]
Test passing (failures are transient)
image

**Comment by kosinsky** [comment]
Looks good for me. I ran some acceptance tests (new and *_basic)  they are green too.

**Comment by jackofallops** [comment]
> @jackofallops, Looks good for me. I ran some acceptance tests (new and *_basic) they are green too.

Brilliant, thanks!  I'll merge this then follow up on your other PR for monitor next.

**Pull request merged**